src/app/AttributeCache.h

```diff
@@ -0,0 +1,368 @@
+        {
+            return CHIP_ERROR_INVALID_ARGUMENT;
+        }
+
+        status = attributeState->Get<StatusIB>();
+        return CHIP_NO_ERROR;
+    }
+
+    /*
+     * Encapsulates a StatusIB and a ConcreteAttributePath pair.
+     */
+    struct AttributeStatus
+    {
+        AttributeStatus(ConcreteAttributePath & path, StatusIB & status) : mPath(path), mStatus(status) {}
+        ConcreteAttributePath mPath;
+        StatusIB mStatus;
+    };
+
+    /*
+     * Retrieve the value of an entire cluster instance from the cache (if present) given a path
+     * and decode it using DataModel::Decode into the in-out argument 'value'. If any StatusIBs
+     * are present in the cache instead of data, they will be provided in the statusList argument.
+     *
+     * The template parameter ClusterObjectT is generally expected to be a
+     * ClusterName::Attributes::DecodableType, but any
+     * object that can be decoded using the DataModel::Decode machinery will work.
+     *
+     * Notable return values:
+     *      - If neither data or status for the specified path don't exist in the cache, CHIP_ERROR_KEY_NOT_FOUND
+     *        shall be returned.
+     *
+     */
+    template <typename ClusterObjectTypeT>
+    CHIP_ERROR Get(EndpointId endpointId, ClusterId clusterId, ClusterObjectTypeT & value, std::list<AttributeStatus> & statusList)
+    {
+        statusList.clear();
+
+        return ForEachAttribute(endpointId, clusterId, [&value, this, &statusList](const ConcreteAttributePath & path) {
+            TLV::TLVReader reader;
+            CHIP_ERROR err;
+
+            err = Get(path, reader);
+            if (err == CHIP_ERROR_IM_STATUS_CODE_RECEIVED)
+            {
+                ConcreteAttributePath pathCopy(path);
+                StatusIB status;
+
+                ReturnErrorOnFailure(GetStatus(path, status));
+                statusList.push_back(AttributeStatus(pathCopy, status));
+                err = CHIP_NO_ERROR;
+            }
+            else if (err == CHIP_NO_ERROR)
+            {
+                ReturnErrorOnFailure(DataModel::Decode(reader, path, value));
+            }
+            else
+            {
+                return err;
+            }
+
+            return CHIP_NO_ERROR;
+        });
+    }
+
+    /*
+     * Retrieve the value of an attribute by updating a in-out TLVReader to be positioned
+     * right at the attribute value.
+     *
+     * Notable return values:
+     *      - If neither data nor status for the specified path don't exist in the cache, CHIP_ERROR_KEY_NOT_FOUND
+     *        shall be returned.
+     *
+     *      - If a StatusIB is present in the cache instead of data, a CHIP_ERROR_IM_STATUS_CODE_RECEIVED error
+     *        shall be returned from this call instead. The actual StatusIB can be retrieved using the GetStatus() API above.
+     *
+     */
+    CHIP_ERROR Get(const ConcreteAttributePath & path, TLV::TLVReader & reader);
```

Comment:
Should document that this reader only remains valid until the cached value for the path is updated, so must not be held across any async work....

Reply:
I actually have to add that note to the other `Get`s too, that the memory backing the DecodableType (especially for lists and octet strings) will disappear from underneath you across async work.

If this API gets utilized enough, I have plans to add an ability to retain the underlying buffers as a handle/object that allows you to then hold on to the decoded values for longer than the current execution context.